From cac0d8bebc269f5f82cf146656f4cb0ef8e6bc49 Mon Sep 17 00:00:00 2001
From: Matthew Schwartz <matthew.schwartz@linux.dev>
Date: Sat, 6 Sep 2025 21:02:30 -0700
Subject: [PATCH 31/49] hid-asus-ally: disable wakeup attribute on N-Key device

With the current ROG Ally patchset, the MCU will send spurious events
when the ROG Ally is in s2idle while plugged in, which can cause the
system to wake unexpectedly.

There's really no reason the N-Key needs wakeup enabled, so just
disable the capability for now. This can be dropped when the issue is
root caused and fixed within the standard ROG Ally patchset.

Closes: https://github.com/ValveSoftware/SteamOS/issues/2119
Signed-off-by: Matthew Schwartz <matthew.schwartz@linux.dev>
---
 drivers/hid/hid-asus-ally.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/drivers/hid/hid-asus-ally.c b/drivers/hid/hid-asus-ally.c
index ae3f257c7..8b51560fd 100644
--- a/drivers/hid/hid-asus-ally.c
+++ b/drivers/hid/hid-asus-ally.c
@@ -2010,6 +2010,23 @@ static int ally_hid_init(struct hid_device *hdev)
 	return ret;
 }
 
+static void ally_disable_nkey_wakeup(struct hid_device *hdev)
+{
+	struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
+	struct usb_device *udev;
+
+	if (!intf)
+		return;
+
+	udev = interface_to_usbdev(intf);
+	if (!udev)
+		return;
+
+	/* HACK: Mark Ally N-Key as incapable of wakeup */
+	device_set_wakeup_enable(&udev->dev, false);
+	hid_info(hdev, "Disabled wakeup capability on %s\n", dev_name(&udev->dev));
+}
+
 static int ally_hid_probe(struct hid_device *hdev, const struct hid_device_id *_id)
 {
 	struct usb_interface *intf = to_usb_interface(hdev->dev.parent);
@@ -2054,6 +2071,7 @@ static int ally_hid_probe(struct hid_device *hdev, const struct hid_device_id *_
 	/* This should almost always exist */
 	if (ep == ROG_ALLY_CFG_INTF_IN) {
 		validate_mcu_fw_version(hdev, idProduct);
+		ally_disable_nkey_wakeup(hdev);
 
 		drvdata.led_rgb_dev = ally_rgb_create(hdev);
 		if (IS_ERR(drvdata.led_rgb_dev))
-- 
2.52.0

